daemons/attrd/attrd_sync.c

```diff
@@ -200,8 +327,245 @@ attrd_request_sync_point(xmlNode *xml)
     }
 }
 
+/*!
+ * \internal
+ * \brief Does an IPC request contain any sync point attribute?
+ *
+ * \param[in] xml   An XML IPC request
+ *
+ * \return true if there's a sync point attribute, false otherwise
+ */
 bool
 attrd_request_has_sync_point(xmlNode *xml)
 {
     return attrd_request_sync_point(xml) != NULL;
 }
+
+static void
+free_action(gpointer data)
+{
+    struct confirmation_action *action = (struct confirmation_action *) data;
+    g_list_free_full(action->respondents, free);
+    mainloop_timer_del(action->timer);
+    free_xml(action->xml);
+    free(action->client_id);
+    free(action);
+}
+
+/* Remove an IPC request from the expected_confirmations table if the peer attrds
+ * don't respond before the timeout is hit.  We set the timeout to 15s.  The exact
+ * number isn't critical - we just want to make sure that the table eventually gets
+ * cleared of things that didn't complete.
+ */
+static gboolean
+confirmation_timeout_cb(gpointer data)
+{
+    struct confirmation_action *action = (struct confirmation_action *) data;
+
+    GHashTableIter iter;
+    gpointer value;
+
+    if (expected_confirmations == NULL) {
+        return G_SOURCE_REMOVE;
+    }
+
+    g_hash_table_iter_init(&iter, expected_confirmations);
+
+    while (g_hash_table_iter_next(&iter, NULL, &value)) {
+        if (value == action) {
+            pcmk__client_t *client = pcmk__find_client_by_id(action->client_id);
+            if (client == NULL) {
+                return G_SOURCE_REMOVE;
+            }
+
+            crm_trace("Timed out waiting for confirmations for client %s", client->id);
+            pcmk__ipc_send_ack(client, action->ipc_id, action->flags | crm_ipc_client_response,
+                               "ack", ATTRD_PROTOCOL_VERSION, CRM_EX_TIMEOUT);
+
+            g_hash_table_iter_remove(&iter);
+            crm_trace("%d requests now in expected confirmations table", g_hash_table_size(expected_confirmations));
+            break;
+        }
+    }
+
+    return G_SOURCE_REMOVE;
+}
+
+/*!
+ * \internal
+ * \brief When a peer disconnects from the cluster, no longer wait for its confirmation
+ *        for any IPC action.  If this peer is the last one being waited on, this will
+ *        trigger the confirmation action.
+ *
+ * \param[in] host   The disconnecting peer attrd's uname
+ */
+void
+attrd_do_not_expect_from_peer(const char *host)
+{
+    GList *keys = g_hash_table_get_keys(expected_confirmations);
```

Comment:
We need a NULL check here. During some shutdowns (maybe more reliable when the cluster is in the process of initially starting resources?):
```
Nov 15 02:47:48.299 fastvm-rhel-9.0-42 pacemaker-attrd     [1293] (crm_glib_handler)    crit: GLib: g_hash_table_get_keys: assertion 'hash_table != NULL' failed
Nov 15 02:47:48.299 fastvm-rhel-9.0-42 pacemaker-attrd     [1293] (update_peer_state_iter)      notice: Node fastvm-rhel-9.0-41 state is now lost | nodeid=1 previous=member source=crm_update_peer_proc
Nov 15 02:47:48.299 fastvm-rhel-9.0-42 pacemaker-attrd     [1293] (attrd_peer_remove)   notice: Removing all fastvm-rhel-9.0-41 attributes for peer loss
Nov 15 02:47:48.299 fastvm-rhel-9.0-42 pacemaker-attrd     [1293] (fail_assert_as)      error: crm_glib_handler: Forked child [1849] to record non-fatal assertion at logging.c:75 : g_hash_table_get_keys: assertion 'hash_table != NULL' failed
Nov 15 02:47:48.367 fastvm-rhel-9.0-42 pacemaker-attrd     [1293] (crm_glib_handler)    crit: GLib: g_hash_table_get_keys: assertion 'hash_table != NULL' failed
```

Reply:
https://github.com/ClusterLabs/pacemaker/pull/2937